### PR TITLE
feat(pipelined): Add IPv6 GTP support to OVS

### PR DESCRIPTION
**Pull request by koolzz**
Signed-off-by: Nick Yurchenko <koolzz@fb.com>

## Summary
```
set /bin/bash './tests/system-kmod-testsuite' -C tests  AUTOTEST_PATH='utilities:vswitchd:ovsdb:vtep:tests::'; \
"$@" 140-154 -j1 || (test X'' = Xyes && "$@" --recheck)
## ------------------------------ ##
## openvswitch 2.15.2 test suite. ##
## ------------------------------ ##

layer3-tunnels

140: layer3 - ping over Bareudp                      skipped (system-layer3-tunnels.at:206)
141: layer3 - ping over GTP                          ok
142: layer3 - IPv6 packets over IPv6 GTP             ok
143: layer3 - IPv6 packets over IPv4 GTP             ok
144: layer3 - IPv4 packets over IPv6 GTP             ok
145: layer3 - GTP echo match test                    ok
146: layer3 - IPv6 GTP echo match test               ok
147: layer3 - GTP echo response test                 ok
148: layer3 - IPv6 GTP echo response test            ok
149: layer3 - GTP echo response test multi endpoint  ok
150: layer3 - IPv6 GTP echo response test multi endpoint ok
151: layer3 - GTP end marker test                    ok
152: layer3 - IPv6 GTP end marker test               ok

system-inteface

153: interface - add delete add same interface       ok
154: interface - add route to br and verify clean-up ok

## ------------- ##
## Test results. ##
## ------------- ##

14 tests were successful.
1 test was skipped.
make[1]: Leaving directory '/root/q/ovs'
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0019-Add-GTP-IPv6-support.patch [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
++ n_packets=1, n_bytes=82, priority=100,ipv6,ipv6_dst=2001::2 actions=drop
++])
++
++dnl sleep 1000
+ OVS_TRAFFIC_VSWITCHD_STOP
+ AT_CLEANUP
+ 
+-- 
+2.25.1
+
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0019-Add-GTP-IPv6-support.patch]: can you add equivalent IPv6 test for all IPv4 GTP tests in this unit test.

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+ 
+-	setup_udp_tunnel_sock(net, sock1u, &tunnel_cfg);
++static void gtp_setup_sock(struct gtp_dev *gtp,
++                                     struct net *net,
++                                     struct socket *sock)
++{
++	struct udp_tunnel_sock_cfg tunnel_cfg;
+ 
+-	sock_hold(sock1u->sk);
++    memset(&tunnel_cfg, 0, sizeof(tunnel_cfg));
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch]: there are indentation errors, can you use tab as indentation in ur editor.

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
++    }
++    udp_conf.local_ip.s_addr = htonl(INADDR_ANY);
++    udp_conf.local_udp_port = htons(GTP1U_PORT);
+ 
+-	udp_conf.family = AF_INET;
+-	udp_conf.local_ip.s_addr = htonl(INADDR_ANY);
+-	udp_conf.local_udp_port = htons(GTP1U_PORT);
++    err = udp_sock_create(net, &udp_conf, &sock);
++    if (err < 0)
++        return err;
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch]: Return type is pointer, can you use ERR_PTR to return error.

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
++
++AT_CHECK([ovs-ofctl dump-flows br0 | ofctl_strip | grep ipv6 ], [0], [dnl
++ priority=100,ipv6,ipv6_dst=2001::1 actions=output:1
++ n_packets=1, n_bytes=63, priority=100,ipv6,ipv6_dst=2001::2 actions=drop
++])
++
++NS_CHECK_EXEC([at_ns0], [tcpdump -l -n -xx -U ip6 host fc00::55:0:111 -i p0 > ipv6_xmit.pcap  2>/dev/null &])
++
++sleep 1
++AT_CHECK([ping -6 2001::1 -q -c 3 -I br0 2>/dev/null | FORMAT_PING], [0], [dnl
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch:938]: can you also add test for IPv4 ping for br0 ip address.

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
++3 packets transmitted, 0 received, 100% packet loss, time 0ms
++])
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch:940]: can you also add test in IPv6 ping for IPv4 tunnel in existing test.

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch [hunk 6]
```diff
@@ -0,0 +1,1405 @@
+From f93c16b78fa4fcb952a2f45bb9b4928f6773d64d Mon Sep 17 00:00:00 2001
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch]: can you fix following warning:
```
/root/work/repo/ovs/datapath/linux/gtp.c: In function ‘gtp_create_sock’:
/root/work/repo/ovs/datapath/linux/gtp.c:701:10: warning: returning ‘int’ from a function with return type ‘struct socket *’ makes pointer from integer without a cast [-Wint-conversion]
  701 |   return err;
      |          ^~~

```

File: third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch [hunk 7]
```diff
@@ -0,0 +1,1405 @@
...
++ n_packets=3, n_bytes=354, priority=100,ipv6,ipv6_dst=2001::1 actions=output:1
++ n_packets=1, n_bytes=63, priority=100,ipv6,ipv6_dst=2001::2 actions=drop
++])
++
++dnl sleep 1000
++OVS_TRAFFIC_VSWITCHD_STOP
++AT_CLEANUP
++
++
++AT_SETUP([layer3 - GTP IPv4 packets over IPv6 tunnel])
```
Comment (pshelar) [inline third_party/gtp_ovs/ovs-gtp-patches/2.15/0022-Add-GTP-IPv6-support.patch]: can you change name to be consistent with above test case to 'IPv4 packets over IPv6 GTP'

**Review by pshelar — COMMENTED** [review]
This looks pretty close.

**Comment by koolzz** [comment]
@pshelar fixed requested changes

**Review by pshelar — APPROVED** [review]
LGTM

**Pull request merged**